designs/nodepool-requirement-flexibility.md

```diff
@@ -0,0 +1,185 @@
+
+## Solutions
+
+### [Recommended] Solution 1: `minValues` field in the `spec.requirements`  section of NodeClaim Template (becomes part of the template on the NodePool)
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+      - key: karpenter.kwok.sh.instance-family
+        operator: In
+        values: ["c", "m", "r"]
+        minValues: 2
+      - key: node.kubernetes.io/instance-type
+        operator: Exists
+        minValues: 10
+```
+
+We would add a new `minValues` field to the requirements section of the NodeClaim template. By adding this section to the NodeClaim template, we inform the scheduler that it should stop continuing to bin-pack pods onto Preflight NodeClaims and should, instead, create another Preflight NodeClaim that will schedule that pod.
+
+#### Pros/Cons
+
+1. 👍👍 Puts all core scheduling behavior together in the same block, allowing users to easily reason about what their requirements and minimums mean
+2. 👎 Piggy-backs off of a well-known core API. There’s risk that if this API changes in a way that is incompatible with this value in the future, we may have to break our `minValues` API
+
+### Solution 2: `minValues` in NodeClass API
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: EC2NodeClass
+spec:
+  minValues:
+    node.kubernetes.io/instance-type: 10
+    karpenter.kwok.sh.instance-family: 2
+---
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+      - key: karpenter.kwok.sh.instance-family
+        operator: In
+        values: ["c", "m", "r"]
+      - key: node.kubernetes.io/instance-type
+        operator: Exists
+```
+
+Enabling this solution allows specific Cloud Providers to make this change without pushing this new field down into the neutral API; however, this means we need to define a neutral way where a cloudprovider can define its own flexibility requirements and pass them back to the scheduler. One option here is to allow CloudProviders to layer on their own flexibility requirements for NodePools through the [CloudProvider interface](https://github.com/kubernetes-sigs/karpenter/blob/4e85912c81ed9a51fc8ebd107db8e88e7828fae7/pkg/cloudprovider/types.go). This function in the interface might look like `func GetFlexibility(np *v1beta1.NodePool) scheduling.Flexibility` where `scheduling.Flexibility` is a `map[string]int`.
+
+#### Pros/Cons
+
+1. 👍👍 Isolates the minValues feature to whether a Cloud Provider wants to implement it or not
+2. 👍 Enables Cloud Providers to have the concept of “default” minValues that is controlled at runtime, allowing cloud providers to enforce certain minimums from their users for things like spot best practices without having to have the user specify it directly (e.g. “don’t allow a spot instance to launch without at least 10 spot instances in the request”).
```

Comment:
also sharing a nodeclass across all nodepools makes it easy to have defaults just applied. Configure it once and share that nodeclass everywhere if you aren't too picky about some of the other settings. 

The downside is nodeclass also provides a lot of settings that make sense to be different. I want multiple image families for example. So sharing one default value across in many places with nodeclass becomes tricky

Reply:
I could see this potentially being hierarchical here. If we ever felt a need for global configuration we could always say that that could be configured and if you configure anything more granular then that overrides the global configuration.

I think we've generally pushed back on global configuration, though. Because you can achieve the same thing by just templating out your manifests through IaC and it isn't truly that cumbersome to have to respecify the same data across manifests if you are using something like helm or kustomize.